src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs

```diff
@@ -19,40 +19,40 @@ private static ArraySegment<byte> CreateAsciiByteArraySegment(string text)
             return new ArraySegment<byte>(bytes);
         }
 
-        public static ArraySegment<byte> BeginChunkBytes(int dataCount)
+        public static int BeginChunkBytes(int dataCount, Span<byte> span)
         {
-            var bytes = new byte[10]
-            {
-                _hex[((dataCount >> 0x1c) & 0x0f)],
-                _hex[((dataCount >> 0x18) & 0x0f)],
-                _hex[((dataCount >> 0x14) & 0x0f)],
-                _hex[((dataCount >> 0x10) & 0x0f)],
-                _hex[((dataCount >> 0x0c) & 0x0f)],
-                _hex[((dataCount >> 0x08) & 0x0f)],
-                _hex[((dataCount >> 0x04) & 0x0f)],
-                _hex[((dataCount >> 0x00) & 0x0f)],
-                (byte)'\r',
-                (byte)'\n',
-            };
-
             // Determine the most-significant non-zero nibble
             int total, shift;
-            total = (dataCount > 0xffff) ? 0x10 : 0x00;
-            dataCount >>= total;
-            shift = (dataCount > 0x00ff) ? 0x08 : 0x00;
-            dataCount >>= shift;
+            var count = dataCount;
+            total = (count > 0xffff) ? 0x10 : 0x00;
+            count >>= total;
+            shift = (count > 0x00ff) ? 0x08 : 0x00;
+            count >>= shift;
             total |= shift;
-            total |= (dataCount > 0x000f) ? 0x04 : 0x00;
+            total |= (count > 0x000f) ? 0x04 : 0x00;
 
-            var offset = 7 - (total >> 2);
-            return new ArraySegment<byte>(bytes, offset, 10 - offset);
+            count = (total >> 2) + 3;
+
+            var offset = 0;
+            for (shift = total; shift >= 0; shift -= 4)
+            {
+                span[offset] = _hex[((dataCount >> shift) & 0x0f)];
```

Comment:
Would converting _hex to a ROS help here?

Reply:
You can do the same optimization as for _endChunkBytes  for _hex too, use the static property and create the hex bytes manually.

Reply:
Not sure that would have any benefit?

Reply:
Let's keep this PR limited to removing the allocation.